File: 2025/4/solution.ts
import * as fs from "fs";
import * as path from "path";

const puzzleInput = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n");

function toCoords(s: string): number[] {
  return s.split(",").map(Number);
}

function parseInput(data: string[]): Map<string, number> {
  const papers: Map<string, number> = new Map();
  for (const [i, row] of data.entries()) {
    for (const [j, ch] of row.split("").entries()) {
      if (ch === "@") {
        papers.set([i, j].toString(), 0);
      }
    }
  }
  for (const [k, _] of papers) {
    const coords = toCoords(k);
    for (let r = coords[0] - 1; r <= coords[0] + 1; r++) {
      for (let c = coords[1] - 1; c <= coords[1] + 1; c++) {
        if (r === coords[0] && c === coords[1]) continue;
        const tmp = [r, c].toString();
        if (papers.has(tmp)) {
          papers.set(tmp, papers.get(tmp)! + 1);
        }
      }
    }
  }
  return papers;
}

function removePaper(papers: Map<string, number>): number {
  let res = 0;
  for (const v of papers.values()) {
    if (v < 4) res++;
  }
  return res;
}

function removePaperContinuous(papers: Map<string, number>): number {
  let res = 0;
  while (true) {
    let removed = 0;
    for (const [k, v] of papers.entries()) {
      if (v < 4) {
        removed++;
        papers.delete(k);
        const coords = toCoords(k);
        for (let r = coords[0] - 1; r <= coords[0] + 1; r++) {
          for (let c = coords[1] - 1; c <= coords[1] + 1; c++) {
            if (r === coords[0] && c === coords[1]) continue;
            const tmp = [r, c].toString();
            if (papers.has(tmp)) {
              papers.set(tmp, papers.get(tmp)! - 1);
            }
          }
        }
      }
    }
    if (removed === 0) break;
    res += removed;
  }
  return res;
}

function part1(papers: Map<string, number>): number {
  return removePaper(papers);
}

function part2(papers: Map<string, number>): number {
  return removePaperContinuous(papers);
}

const paperMap = parseInput(puzzleInput);

console.log(part1(paperMap));
console.log(part2(paperMap));

File: 2025/4/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(puzzleInput))
	fmt.Println(part2(puzzleInput))
}

func parseInput(data []string) map[[2]int]int {
	papers := make(map[[2]int]int)
	for i, row := range data {
		for j, ch := range row {
			if ch == '@' {
				papers[[2]int{i, j}] = 0
			}
		}
	}
	for coord := range papers {
		a, b := coord[0], coord[1]
		for r := a - 1; r <= a+1; r++ {
			for c := b - 1; c <= b+1; c++ {
				if r == a && c == b {
					continue
				}
				tmp := [2]int{r, c}
				if _, ok := papers[tmp]; ok {
					papers[tmp] += 1
				}
			}
		}
	}
	return papers
}

func removePaper(papers map[[2]int]int) int {
	removed := 0
	for _, ct := range papers {
		if ct < 4 {
			removed += 1
		}
	}
	return removed
}

func removePaperContinuous(papers map[[2]int]int) int {
	res := 0
	for {
		removed := 0
		for coord, ct := range papers {
			if ct < 4 {
				removed += 1
				delete(papers, coord)
				a, b := coord[0], coord[1]
				for r := a - 1; r <= a+1; r++ {
					for c := b - 1; c <= b+1; c++ {
						if r == a && c == b {
							continue
						}
						tmp := [2]int{r, c}
						if _, ok := papers[tmp]; ok {
							papers[tmp] -= 1
						}
					}
				}
			}
		}
		if removed == 0 {
			break
		}
		res += removed
	}
	return res
}

func part1(puzzleInput map[[2]int]int) int {
	return removePaper(puzzleInput)
}

func part2(puzzleInput map[[2]int]int) int {
	return removePaperContinuous(puzzleInput)
}
